File: Day 4/second_part.py

from typing import List

def load_grid_from_file(file_name: str) -> List[List[int]]:
    with open(file_name) as f:
        lines = f.read().splitlines()
    grid = [[
        1 if c == '@' else 0 for c in line
    ] for line in lines]

    return grid

def count_and_remove_accessible_rolls(grid):
    accessible_rolls = 0
    grid_copy = [row[:] for row in grid]
    for row in range(0, len(grid)):
        for col in range(0, len(grid[row])):
            if grid[row][col] == 1:
                count = count_neigbours(grid, row, col) - 1  # Remove the cell itself
                if count < 4:
                    accessible_rolls += 1
                    grid_copy[row][col] = 9

    if accessible_rolls == 0:
        return accessible_rolls

    for row in range(0, len(grid)):
        for col in range(0, len(grid[row])):
            if grid_copy[row][col] == 9:
                grid[row][col] = 0

    return accessible_rolls + count_and_remove_accessible_rolls(grid)

def count_neigbours(grid: List[List[int]], row:int, col:int) -> int:
    """
    Counts the number of neighbours of a cell in the grid. Neighbours are defined as the 8 cells surrounding the cell, including diagonals.
    """
    min_col = max(0, col - 1)
    max_col = min(len(grid[row]) - 1, col + 1)
    min_row = max(0, row - 1)
    max_row = min(len(grid) - 1, row + 1)
    count = 0
    for c in range(min_col, max_col + 1):
        for r in range(min_row, max_row + 1):
            if grid[r][c] == 1:
                count += 1
    return count

grid = load_grid_from_file('input_final.txt')
count_and_remove_accessible_rolls(grid)

File: Day 4/second_part.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

std::vector<std::vector<int>> load_grid_from_file(const std::string& file_name) {
    std::ifstream file(file_name);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file: " + file_name);
    }

    std::vector<std::vector<int>> grid;
    std::string line;

    while (std::getline(file, line)) {
        if (!line.empty() && line.back() == '\r') {
            line.pop_back();
        }

        std::vector<int> row;
        row.reserve(line.size());
        for (char c : line) {
            row.push_back(c == '@' ? 1 : 0);
        }
        grid.push_back(row);
    }

    return grid;
}

int count_neighbors(const std::vector<std::vector<int>>& grid, int row, int col) {
    int min_col = std::max(0, col - 1);
    int max_col = std::min(static_cast<int>(grid[row].size()) - 1, col + 1);
    int min_row = std::max(0, row - 1);
    int max_row = std::min(static_cast<int>(grid.size()) - 1, row + 1);

    int count = 0;
    for (int c = min_col; c <= max_col; ++c) {
        for (int r = min_row; r <= max_row; ++r) {
            if (grid[r][c] == 1) {
                ++count;
            }
        }
    }

    return count;
}

int count_and_remove_accessible_rolls(std::vector<std::vector<int>>& grid) {
    int accessible_rolls = 0;
    std::vector<std::vector<int>> grid_copy = grid;

    for (int row = 0; row < static_cast<int>(grid.size()); ++row) {
        for (int col = 0; col < static_cast<int>(grid[row].size()); ++col) {
            if (grid[row][col] == 1) {
                int count = count_neighbors(grid, row, col) - 1;
                if (count < 4) {
                    ++accessible_rolls;
                    grid_copy[row][col] = 9;
                }
            }
        }
    }

    if (accessible_rolls == 0) {
        return accessible_rolls;
    }

    for (int row = 0; row < static_cast<int>(grid.size()); ++row) {
        for (int col = 0; col < static_cast<int>(grid[row].size()); ++col) {
            if (grid_copy[row][col] == 9) {
                grid[row][col] = 0;
            }
        }
    }

    return accessible_rolls + count_and_remove_accessible_rolls(grid);
}

int main() {
    std::vector<std::vector<int>> grid = load_grid_from_file("input_final.txt");
    int result = count_and_remove_accessible_rolls(grid);
    std::cout << result << std::endl;
    return 0;
}